text/0680-implicit-injection-deprecation.md

```diff
@@ -0,0 +1,127 @@
+
+2. Eager initialization of the dependency tree from which the property came from prevents tree shaking and incurs a
+   performance hit. For example, if you app does not need `ember-data` to render the entry route, the `store` injection will
+   take up more CPU cycles than necessary, hurting common user facing metrics.
+
+3. Eager initialization makes it hard to users to write tests that need to take advantage of a stub.
+   This is often a silent and annoying hurdle that users have to get around when writing tests.
+   Moreover, this is especially apparent for users moving to Ember's new testing APIs in [RFC 232](https://github.com/emberjs/rfcs/blob/master/text/0232-simplify-qunit-testing-api.md) and [RFC 268](https://github.com/emberjs/rfcs/blob/master/text/0268-acceptance-testing-refactor.md),
+   ensuring instance initializers run before each test.
+
+4. Injection on commonly used framework objects can lead to deviation of project specific styleguides. For instance,
+   injecting the `router` on the Ember.Component factory object will certainly lead to instances where the `router` is explicitly
+   injected and other instances where is it implicitly injected.
+
+5. Users have come to associate `Route.model()` with a hook that returns `ember-data` models in the absense of explicit injection.
+   While this can be true, it is not wholly true. New patterns of data loading are becoming
+   accepted in the community including opting to fetch data in a Component.
+
+
+## Detailed design
+
+Removing implicit injections is not possible until we first issue deprecations.
+This helps us maintain our strict semver commitment to not completing a major version bump and
+removing APIs without a deprecation path. As a result, it will take us 3 phases to remove `owner.inject`.
+
+### 1. Deprecate implicit injection on target object
+
+The first step will be issuing a deprecation on the target factory object that is receiving a property. This will surface the deprecation
+on user's owned objects including transitively through addons like `ember-data`. This can be accomplished by installing a native getter
+and setter on the target. Whenever the property is accessed, we will check if the property is explicitly injected. If it isn't,
+we will issue a deprecation with an until value, id and url to read more about this deprecation.
+
+### 2. Deprecate `owner.inject`
+
+The first phase did not actually deprecate the "use" of `owner.inject`.  As a result, we need to deprecate
+it's use directly before removing.
+
+### 3. Profit!
+
+Ember 4.0 will remove the ability to utilize `owner.inject` to inject arbitrary properties on Ember's Framework objects.
+
+It is important to consider the timeline of these three phases.  Many addons and apps will need to make minor and major
+changes to their codebase. We need to consider this as we move through each phase.
```

Comment:
I agree that considering the phases is important, can you update to indicate relative timing (and likely "triggers")? As it stands now, we say that the timing is super important but don't actually say what the timing is or should be 😆

Reply:
On a side note, I may be wrong here because of some misunderstanding or sth.

But I remember that there used to be some feature (that i have never relied on) where an implicit route would automagically load a model, is this a thing or is this something i just made up?

The question is, if it exists, how its going to be deprecated?

Reply:
Is this what you are referring to?

https://github.com/emberjs/rfcs/pull/557
https://github.com/emberjs/rfcs/issues/377

Reply:
Looks like it is it, thank you @snewcomer.
Was wondering how these behaviors play out in this RFC

Reply:
As for the triggers, how does this sound?

1. Introduce deprecation on target
2. Next release introduce deprecation on owner.inject
4. Simmer until v4?